### PR TITLE
Handle bigints in the `TokenAmountInput`

### PR DESCRIPTION
**Pull request by jagodarybacka**
Resolves https://github.com/tahowallet/dapp/issues/694

### Problem

We have been using floating point numbers to transform bigints to strings with 18 decimal places. But it was easy to mess up the precision because bigints were larger than max safe int. We need a way to be able to work in the inputs with both very large and very small numbers.

### What has been done
Make input component handle value changes with bigints not strings. Adjust components using the `TokenAmountInput` to new format.

- show nicely small balances (`<0,0001 TAHO`) - `bigIntToDisplayUserAmount` should be able to display nice label for values smaller than desired precision, let's make it to show the smallest possible number like: `<0.01` for desired precision `2` etc.
- Allow using "Max" button and don't allow the input to override the real value with stringified value with less precision.
- Add util to display bigints as strings with max precision

image

### Testing

To test it please use stake and unstake forms. Type some value or click "max" and try to unstake/stake - check the value that is displayed on the wallet token allowance screen (make sure you don't have any allowance set on the TAHO or veTAHO on that account so you'll be able to test it that way)

- [x] test it with super small amounts (up to 18 decimal places)
- [x] test it with big amounts (for example >=3301 TAHO)
- [x] test it with big amounts with some values on decimal places (`3000.0000000001`)

### CODE REVIEW

**Comment by andreachapman** [comment]
@jagodarybacka - should we do away with these tool tips? I only see them if I hover there for a minute. I don't actually get an error for trying to stake or unstake those amounts. I got those amounts by clicking Max on stake and unstake.
Screenshot 2023-11-15 at 12 29 56 PM
Screenshot 2023-11-15 at 12 30 12 PM

Otherwise, I did the other tests you listed and it seems to be working for me. I only show 4 places after the decimal but when there's greater precision in the actual balance, I can see it and fully unstake or stake it by clicking Max. 

Though I'm also interested in the review from others because I feel like I'm not _as well-versed_ in it as you guys are.

**Comment by jagodarybacka** [comment]
@andreachapman that browser input's tooltip should be hidden now

File: src/shared/components/TokenAmountInput.tsx [hunk 1]
```diff
@@ -46,42 +46,72 @@ export default function TokenAmountInput({
...
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [textAmount, onChange])
+
+  useEffect(() => {
+    // Allow clearing the input from parent componentthis should be the only case
+    // where parent component is allowed to change the value
+    if (amount === null) {
+      setTextAmount("")
+    }
+  }, [amount])
```
Comment (xpaczka) [inline src/shared/components/TokenAmountInput.tsx:100]: I would turn it into a hook f.e. `useTextAmount` - it may be useful in the future

**Review by xpaczka — CHANGES_REQUESTED** [review]
Overally looks good 🚀 tests also successful

File: src/shared/components/TokenAmountInput.tsx (hunk 1 above)
Comment (jagodarybacka) [reply src/shared/components/TokenAmountInput.tsx:100]: Really? I think it is very specific to this component, don't know what component other than input would need it. I would for sure want to battle test it first before making it reusable 😅

**Review by xpaczka — APPROVED** [review]

**Pull request merged**